OpenGL Pipeline

sxzhang25 opened the pull request:
Incorporating OpenGL pipeline (a simple copy shader, no homographies applied)

sxzhang25 committed:
Add ShaderPipeline class.

- Include code to load shaders and create program
- Include methods to link shaders to program

sxzhang25 committed:
Add OpenGL pipeline setup files

- OutputSurface sets up (1) an EGLSurface for rendering based on the provided native surface that is passed to MotionPhotoReader; (2) a surface texture and corresponding surface for the media codec to decode
- TextureRender sets up a renderer to write the image onto the coordinate system (as a texture mapped onto a single rectangular plane represented by two triangles)
- Add basic texture rendering vertex and fragment shaders

sxzhang25 committed:
Add OpenGL pipeline setup files

- OutputSurface sets up (1) an EGLSurface for rendering based on the provided native surface that is passed to MotionPhotoReader; (2) a surface texture and corresponding surface for the media codec to decode
- TextureRender sets up a renderer to write the image onto the coordinate system (as a texture mapped onto a single rectangular plane represented by two triangles)
- Add basic texture rendering vertex and fragment shaders
- Connect OpenGL pipeline to mediacodec handler thread
- Integrate OpenGL pipeline into MotionPhotoReader and BufferProcessor classes: (1) Set lowResDecoder to decode to the surface wrapped around the surface texture generated by an OutputSurface that is connected to an EGL context; (2) Add rendering steps onto render handler thread (previously media handler thread)

sxzhang25 committed:
Give access to render surface from OutputSurface instead of surface texture, and release the surface in OutputSurface release() method.

sxzhang25 committed:
Refactor widget instrumentation tests so they are independent of other activities and instrumentation tests.

- Create a new activity used specifically for motion photo widget tests
- Update Android manifest and new activity layout
- Update main activity to separate it from motion photo widget
- Use new activity in MotionPhotoWidgetTest
- Fix null reader calls in MotionPhotoReader

sxzhang25 committed:
Allow buffer processor to continue with a null surface (ignore output surface calls)

sxzhang25 committed:
Add OutputSurface instrumentation tests and update OutputSurface to include testing capabilities.

sxzhang25 committed:
Update BufferProcessor unit tests to reflect dependency on OutputSurface.

sxzhang25 committed:
Add initial OutputSurface unit tests, along with independent OutputSurface test activity and XML layout files.

sxzhang25 committed:
Update Javadoc.

sxzhang25 committed:
Update contexts for test activity layouts.

sxzhang25 committed:
Small changes (move lines of code around).

sxzhang25 committed:
Update output surface tests.

clju commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/OutputSurface.java:
```diff
@@ -0,0 +1,268 @@
+                    EGL14.EGL_GREEN_SIZE, 8,
+                    EGL14.EGL_BLUE_SIZE, 8,
+                    EGL14.EGL_ALPHA_SIZE, 8,
+                    EGL14.EGL_DEPTH_SIZE, 8,
```
Depth is not needed here.

clju commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/TextureRender.java:
```diff
@@ -0,0 +1,243 @@
+    private static final int TRIANGLE_VERTICES_DATA_UV_OFFSET = 3; // texture coords begin 3 offset units from array start (skip the position coordinates)
+
+    private static final String VERTEX_SHADER =
+            "uniform mat4 u_Matrix;\n" +
```
Let's try to use the more modern OpenGL ES 3.0.

attribute -> in
varying -> out / in
texture2D -> texture

etc.

clju commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/TextureRender.java:
```diff
@@ -0,0 +1,243 @@
+            "}";
+
+    // A single plane comprised of two triangles to hold the image texture
+    private final float[] triangleVerticesData = {
```
No need to pass that much information to the GPU. The third spatial dimension is unneeded, and uv can be recovered from xy with (1.0 + pos) / 2.0.

sxzhang25 replied:
Changes: moved this calculation inside the vertex shader and only stored (x,y) for each corner.

clju commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/TextureRender.java:
```diff
@@ -0,0 +1,243 @@
+        final int[] linkStatus = new int[1];
+        GLES20.glGetProgramiv(program, GLES20.GL_LINK_STATUS, linkStatus, /* offset = */ 0);
+        if (linkStatus[0] == 0) {
+            // Delete the program if linking failed
```
Throw a RuntimeException in that case with the info log - that will help you debug your shader. Same for the shader compilation status below.

clju commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/TextureRender.java:
```diff
@@ -0,0 +1,243 @@
+    private final float[] triangleVerticesData = {
+            // positions (x,y,z)    // texture coordinates (u,v)
+            -1.0f, -1.0f, 0.0f,     0.0f, 0.0f,  // bottom left
+             1.0f, -1.0f, 0.0f,     1,0f, 0.0f,  // bottom right
```
There is a very sneaky mistake here: this line contains SIX numbers and not FIVE. That caused a bunch of issues in your triangles.

clju commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/TextureRender.java:
```diff
@@ -0,0 +1,243 @@
+        // Set the transform matrix of the surface texture
+        surfaceTexture.getTransformMatrix(uMatrix);
+
+        GLES20.glClearColor(/* red = */ 0.0f, /* green = */ 0.0f, /* blue = */ 0.0f, /* alpha = */ 1.0f);
```
This does not need to be done per-frame.

clju commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/TextureRender.java:
```diff
@@ -0,0 +1,243 @@
+
+        GLES20.glUseProgram(program);
+
+        // Process vertices
```
useProgram ; glVertexAttribPointer ; glEnableVertexAttribArray ; glActiveTexture ; glBindTexture ; glUniform1i don't need to be executed per frame. OpenGL is stateful and will remember these operations.

sxzhang25 replied:
Changes: moved to onSurfaceCreated(), after linking the program.

clju commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/TextureRender.java:
```diff
@@ -0,0 +1,243 @@
+        GLES20.glUniform1i(uTextureUnitHandle, /* x = */0);
+
+        GLES20.glDrawArrays(GLES20.GL_TRIANGLE_STRIP, /* first = */ 0, /* count = */ 4);
+        GLES20.glFinish();
```
Is there a reason why you use glFinish here? This is generally not needed / not a good practice.

sxzhang25 replied:
Changes: deleted

clju commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/TextureRender.java:
```diff
@@ -0,0 +1,243 @@
+                /* type = */ GLES20.GL_FLOAT,
+                /* normalized = */ false,
+                TRIANGLE_VERTICES_DATA_STRIDE_BYTES,
+                triangleVertices
```
slightly nicer way to put that is to write trianglesVertices.pos(SOME_OFFSET) in the vertexAttribPointer argument.

clju commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/TextureRender.java:
```diff
@@ -0,0 +1,243 @@
+
+        // Process texture
+        triangleVertices.position(TRIANGLE_VERTICES_DATA_UV_OFFSET);
+        GLES20.glVertexAttribPointer(
```
It would be nicer to use VAOs / VBOs to only upload the data on the GPU once ; but we can work on that later.

sxzhang25 replied:
Added TODO.

clju commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/OutputSurface.java:
```diff
@@ -0,0 +1,268 @@
+import java.util.concurrent.ExecutionException;
+
+/**
+ * Holds state associated with a Surface used for MediaCodec decoder input. Creates an EGL surface
```
nit: one line, blank line, then longer explanation after a <p> symbol.

clju commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/OutputSurface.java:
```diff
@@ -0,0 +1,268 @@
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.KITKAT)
+    public void setSurface(Surface surface) {
```
you need to pass down the size of the Surface too, to set the correct viewport in OpenGL.

clju commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/OutputSurface.java:
```diff
@@ -0,0 +1,268 @@
+    public void onFrameAvailable(SurfaceTexture surfaceTexture) {
+        Log.d(TAG, "New frame available");
+        synchronized (frameSyncObject) {
+            if (frameAvailable) {
```
any reason why you don't directly call drawImage here?

sxzhang25 replied:
Calling the method directly in onFrameAvaialble doesn't work, but I haven't figured out why yet. I'll make a note to look into this more in a separate CL.

clju commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/BufferProcessor.java:
```diff
@@ -123,7 +133,7 @@ public void process(Bundle messageData) {
         Bundle bufferData;
         int bufferIndex;
-        long timestamp;
+        long timestampUs;
```
[commenting here since I can't comment above]

Why do we have a generic process with a Bundle argument? That is generally a pattern to be avoided, could you not have two methods nextFrame and seekTo instead?

sxzhang25 replied:
These are called from the nextFrame and seekTo methods, but I'll add a note to go back and potentially refactor on a separate CL.

sxzhang25 committed:
Second pass of OpenGL pipeline changes.

- Switch to OpenGL 3.0 shader language
- Reduce information stored on GPU (no texture coordinates, only x and y positional coordinates)
- Move setting pointers to vertex data from drawFrame() to onSurfaceCreated() so that these methods are not called per frame
- Set up glViewport and pass the appropriate width and height

sxzhang25 committed:
Change static open() call to include arguments for surface dimensions.

sxzhang25 committed:
Update OpenGL pipeline to render video to surface.

- Connect renderer to texture from decoder surface texture.
- Resize the video so that it fits the surface exactly.

sxzhang25 committed:
Fix unit tests to update changes in MotionPhotoReader.

sxzhang25 committed:
Merge branch 'master' of https://github.com/googleinterns/motion-photos-filereader into open-gl

Update from master.

# Conflicts:
#	MotionPhotoReader/app/src/androidTest/java/com/google/android/libraries/motionphotoreader/MotionPhotoWidgetTest.java
#	MotionPhotoReader/app/src/main/AndroidManifest.xml
#	MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoReader.java
#	MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/WidgetTestActivity.java
#	MotionPhotoReader/app/src/main/res/layout/activity_widget_test.xml

sxzhang25 committed:
Refactor all methods to read raw/resource/asset files into a new class. Update all unit and instrumentation tests to use methods from this class.

sxzhang25 committed:
Fix setFile() method in MotionPhotoWidget and update all Javadoc.

sxzhang25 committed:
Update visual anchoring of widget.

- Add support for video fill mode (scale to fit or scale to fill).

sxzhang25 committed:
Set up background color support in OpenGL pipeline.

sxzhang25 committed:
Remove "fill" attribute.

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/ResourceFetcher.java:
```diff
@@ -0,0 +1,83 @@
+package com.google.android.libraries.motionphotoreader;
```
can we move this to the javatests/com/google/android/libraries/motionphotoreader  folder? This way, we don't accidentally use this in regular code

sxzhang25 replied:
I use this in both instrumentation tests and unit tests (and also the widget test activity) so moving it to one of the test folders makes the activity and the other type of test files lose access to it. Is there another way I should organize this class?

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoReader.java:
```diff
@@ -147,6 +229,7 @@ private void startRenderThread(MotionPhotoInfo motionPhotoInfo) throws IOExcepti
         lowResExtractor.setDataSource(fd, file.length() - videoOffset, videoOffset);
 
         // Find the video track and create an appropriate media decoder
+        // TODO: Initialize high resolution media decoder (for stabilization)
```
we won't need to stabilize the high-res track, so won't need a high-res mediadecoder.

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/OutputSurface.java:
```diff
@@ -0,0 +1,288 @@
+     */
+    public Surface getDecodeSurface() {
+        SettableFuture<Surface> surfaceFuture = SettableFuture.create();
+        renderHandler.post(() -> surfaceFuture.set(decodeSurface));
```
couldn't this get set with null ?

sxzhang25 replied:
If I synchronize the block where the decodeSurface is created, would that fix this? My intention is that this would only return null if getDecodeSurface() was called before setSurface() was called (where I initialize decodeSurface), and I would put a note of not to do in the method javadoc.

sxzhang25 committed:
Extract stabilization matrices and apply stabilization.

- Integrate protobuf to get stabilization data from motion track for v1 photos.
- Incorporate stabilization matrices into OpenGL pipeline for a single triangle strip.

Signed-off-by: sxzhang25 <zsharonx@gmail.com>

sxzhang25 committed:
Move decodeSurface initialization to avoid unintentionally returning a null decode surface.

sxzhang25 committed:
Synchronize decodeSurface initialization to avoid unintentionally returning a null decode surface.

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/OutputSurface.java:
```diff
@@ -0,0 +1,288 @@
+            surfaceTexture = new SurfaceTexture(surfaceTextureHandle);
+            surfaceTexture.setOnFrameAvailableListener(this);
+            decodeSurface = new Surface(surfaceTexture);
+        });
```
make decodeSurface a settablefuture and set it here.

sxzhang25 committed:
Add Protobuf files.

rjuang commented on MotionPhotoReader/lib/src/main/java/com/example/lib/Protobuf.java:
```diff
@@ -0,0 +1,4 @@
+package com.example.lib;
+
```
revert this file?

rjuang commented on MotionPhotoReader/app/src/androidTest/java/com/google/android/libraries/motionphotoreader/OutputSurfaceTest.java:
```diff
@@ -0,0 +1,137 @@
+    private static final long SEEK_AMOUNT_US = 10_000L;
+    private final static int WIDTH = 4032;
+    private final static int HEIGHT = 3024;
+
```
nit: remove extra blank line

rjuang commented on MotionPhotoReader/app/src/androidTest/java/com/google/android/libraries/motionphotoreader/OutputSurfaceTest.java:
```diff
@@ -0,0 +1,137 @@
+        cleanup.clear();
+    }
+
+    private File fetchAssetFile(String filename) throws IOException {
+        InputStream input = InstrumentationRegistry.getInstrumentation()
+                .getContext()
+                .getResources()
+                .getAssets()
+                .open(filename);
+
+        // Write file to temporary folder for instrumentation test access
+        File f = temporaryFolder.newFile(filename);
+        writeBytesToFile(input, f);
+        return f;
+    }
```
isn't this in ResourceFetcher now ?

rjuang commented on MotionPhotoReader/app/src/androidTest/java/com/google/android/libraries/motionphotoreader/OutputSurfaceTest.java:
```diff
@@ -0,0 +1,137 @@
+        return f;
+    }
+
+    private static void writeBytesToFile(InputStream input, File file) throws IOException {
+        try (FileOutputStream fos = new FileOutputStream(file)) {
+            ByteStreams.copy(input, fos);
+        }
+    }
+
```
wasn't this also in ResourceFetcher ? .. is it even used?

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoReader.java:
```diff
@@ -62,22 +64,61 @@
+     */
+    @RequiresApi(api = M)
+    public static MotionPhotoReader open(
+            File file,
+            Surface surface,
+            int surfaceWidth,
+            int surfaceHeight
```
indentation looks weird, also move the ) to same line as last input argument

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoReader.java:
```diff
@@ -131,7 +208,12 @@ static MotionPhotoReader open(File file,
+     * extractors.
+     *
+     * A low resolution decoder and extractor are set up for the video track which holds frame data.
+     * If applicable, a high resolution decoder and extractor are set up for the motion track, which
+     * contains video stabilization data.
```
what do you mean by high-resolution decoder? We are only decoding the low-res track and not the moments track.

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoWidget.java:
```diff
@@ -100,19 +101,21 @@ public void surfaceCreated(SurfaceHolder holder) {
                 }
                 try {
-                    reader = MotionPhotoReader.open(file, holder.getSurface());
+                    reader = MotionPhotoReader.open(file, holder.getSurface(), surfaceWidth, surfaceHeight);
```
nit: line break so that fits within 100-char limit

rjuang commented on MotionPhotoReader/app/src/test/java/com/google/android/libraries/motionphotoreader/BufferProcessorTest.java:
```diff
@@ -53,10 +73,48 @@ public void setUp() {
+                                              int rotation,
+                                              String mime) {
+        MediaFormat videoFormat = mock(MediaFormat.class);
+        doAnswer((Answer<Integer>) invocation -> width)
+                .when(videoFormat)
+                .getInteger(eq(MediaFormat.KEY_WIDTH));
+        doAnswer((Answer<Integer>) invocation -> height)
+                .when(videoFormat)
+                .getInteger(eq(MediaFormat.KEY_HEIGHT));
+        doAnswer((Answer<Long>) invocation -> duration)
+                .when(videoFormat)
+                .getLong(eq(MediaFormat.KEY_DURATION));
+        doAnswer((Answer<Integer>) invocation -> rotation)
+                .when(videoFormat)
+                .getInteger(eq(MediaFormat.KEY_ROTATION));
+        doAnswer((Answer<String>) invocation -> mime)
+                .when(videoFormat)
+                .getString(eq(MediaFormat.KEY_MIME));
+
```
doAnswer for method calls that need a side effect other than returning values. For these cases, it'll be easier to read as:

```
when(videoFormat.getString(MediaFormat.KEY_WIDTH)).thenReturn(width);
when(videoFormat.getString(MediaFormat.KEY_HEIGHT)).thenReturn(height);
...  
when(videoFormat.getString(MediaFormat.KEY_MIME)).thenReturn(mime);
```

Note eq(...) is implied and isn't needed unless one of the other args in the method call is a pattern matcher

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/ResourceFetcher.java:
```diff
@@ -0,0 +1,83 @@
+
+import com.google.common.io.ByteStreams;
+
+import org.junit.rules.TemporaryFolder;
```
organize imports.. junit rules shouldn't be used here.

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/ResourceFetcher.java:
```diff
@@ -0,0 +1,83 @@
+            ByteStreams.copy(input, fos);
+        }
+    }
+
```
nit: remove blank line

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/ResourceFetcher.java:
```diff
@@ -0,0 +1,83 @@
+    public static File fetchAssetFile(Context context,
+                                      TemporaryFolder temporaryFolder,
+                                      String filename) throws IOException {
+        InputStream input = context.getResources().getAssets().open(filename);
+
+        // Write file to temporary folder for instrumentation test access
+        File f = temporaryFolder.newFile(filename);
+        writeBytesToFile(input, f);
```
input isn't closed. use try-with-Resources block:

```
try (InputStream input = context.getResources().getAssets().open(filename)) {
   File f = ...
   writeBytesToFile ...
   return f;
}
```

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/ResourceFetcher.java:
```diff
@@ -0,0 +1,83 @@
+    public static File fetchRawFile(Context context,
+                                    int id, String prefix,
+                                    String suffix) throws IOException {
+        InputStream input = context.getResources().openRawResource(id);
+        File file = null;
```
input stream leaked

sxzhang25 committed:
Clean up code.

- Refactor ResourceFetcher methods so that the class does not import JUnit methods.
- Rewrite instrumentation tests to use when(...).thenReturn(...) instead of doAnswer(...).when(...).
- Optimize import statements.
- Fix code formatting.

rjuang commented on MotionPhotoReader/app/src/androidTest/java/com/google/android/libraries/motionphotoreader/OutputSurfaceTest.java:
```diff
@@ -0,0 +1,103 @@
+ */
+@RunWith(AndroidJUnit4.class)
+public class OutputSurfaceTest {
+    private static final String filename = "MVIMG_20200621_200240.jpg";
+
```
static final variables are always named with all caps to indicate they are constants. So, rename as FILENAME, or better yet TEST_VIDEO_FILENAME.

sxzhang25 committed:
Fix tests.

sxzhang25 committed:
Upload additional v1 test files.

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/OutputSurface.java:
```diff
@@ -0,0 +1,294 @@
+
+    private int surfaceTextureHandle;
+    private SurfaceTexture surfaceTexture;
+    private SettableFuture<Surface> decodeSurfaceFuture = SettableFuture.create();
```
set final

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/OutputSurface.java:
```diff
@@ -0,0 +1,294 @@
+    private SurfaceTexture surfaceTexture;
+    private SettableFuture<Surface> decodeSurfaceFuture = SettableFuture.create();
+    private TextureRender textureRender;
+    private Handler renderHandler;
```
this can be final too

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/OutputSurface.java:
```diff
@@ -0,0 +1,294 @@
+    private TextureRender textureRender;
+    private Handler renderHandler;
+    private boolean frameAvailable;
+    private MotionPhotoInfo motionPhotoInfo;
```
this can be final

sxzhang25 committed:
Set constructor variables to final.
